### ISSUE TITLE
Allow RigidBody to have Colliders that are indirect children, not just direct children

### ISSUE THREAD

**Issue by aaronfranke** [godotengine/godot-proposals#535]
**Describe the project you are working on:**

I would like to create a voxel game where the objects' collisions are composed of many cube colliders. I can't use concave mesh colliders because I would like these voxel objects to be able to move.

**Describe the problem or limitation you are having in your project:**

The problem is that, while I can easily generate cube colliders, I have no good way to remove groups of them whenever I need to regenerate a "chunk".

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**

What I would like to do is have several of these cube colliders together as children of a child so that I can easily remove/generate/instance them as needed.

```
RigidBody
    Collider0
    ColliderGroupA
        Collider1
        Collider2
    ColliderGroupB
        Collider3
        Collider4
```

In the example above, in the current stable and master branch of Godot, only Collider0 is detected, and without it the RigidBody complains that it has no collider. It seems that colliders are only used if they are direct children.

Ideally, RigidBody should be able to use colliders that are grandchildren etc and not just direct children. This would make it easier for me to instance, generate, and delete groups of colliders. Basically, the RigidBody should recursively search its children for CollisionShape nodes, so that in the example above, it would use all of the colliders (0 through 4).

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**

One issue that would arise from just recursively searching for collider children is that physics objects parented to physics objects would both use the child's colliders. There are many possible implementations that would solve this without that problem:

* Configurable recursiveness/depth of colliders.

* Don't look for colliders under other PhysicsBody nodes.

* Only use direct children, but add a new "CompoundCollider" node which passes up its own children.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**

Possibly, but not nearly as clean. I really don't want to short-circuit the node hierarchy.

**Is there a reason why this should be core and not an add-on in the asset library?:**

This is a low-level feature that could not be implemented very well by an add-on.

See: https://godotengine.org/qa/31701/possible-rigidbody-have-colliders-that-not-direct-children

**Comment by AndreaCatania** [comment]
Consider that by default we are using compound under the hood, and I prefer to not add a node of this kind to allow recursive search.

A shape may have the possibility to find the physics body owner recursively, even when a physics body contains another one: The shape will stop its search soon it found its owner.

```
Physics body
|--Node
|    |- Shape
|
| -- PhysicsBody 2
|       |--Node
|            |- Shape
```

**Comment by jcarlosrc** [comment]
So, in current implementation a kinematicBody basically disables all collision shapes below, true? Is there a way we can enable them in code?

**Comment by madmiraal** [comment]
> What I would like to do is have several of these cube colliders together as children of a child so that I can easily remove/generate/instance them as needed.
```
RigidBody
    Collider0
    ColliderGroupA
        Collider1
        Collider2
    ColliderGroupB
        Collider3
        Collider4
```
This is how a `CollisionPolygon` currently works. It is a `CollisionObject` that contains multiple shapes (whereas a `CollisionShape` is a `CollisionObject` that just holds one shape). Instead of the above layout, what you actually currently have is:
```
RigidBody
    Collider0
        Shape0
    Collider1
        Shape0
        Shape1
    Collider2
        Shape0
        Shape1
```
You can create your own using the `CollisionObject` [API](https://docs.godotengine.org/en/stable/classes/class_collisionobject.html) API. Create a `shape_owner` using [`create_shape_owner()`](https://docs.godotengine.org/en/stable/classes/class_collisionobject.html#class-collisionobject-method-create-shape-owner), then add multiple shapes to it using [`shape_owner_add_shape()`](https://docs.godotengine.org/en/stable/classes/class_collisionobject.html#class-collisionobject-method-shape-owner-add-shape).

**Comment by AndreaCatania** [comment]
In a recent project, one of the most common behaviors was adding a lot of shapes per single Physics Body; this was causing some performance issues, since the broad phase was not able to properly do its job.

Add 1 shape per Physics Body should be the way to go (unless you want to decompose a complex body shape to multiple convex shapes). This proposal would encourage put more shapes per body even more.

While initially I supported this feature, I think we should reconsider it, so to better guide the user.

**Comment by KevinTriplett** [comment]
@AndreaCatania so if best practice for performance is one shape and one `CollisionObject`, then when @aaronfranke wants to add a group of shapes (shapes from a group of new children) then iterate through the children and add each child's shape using `shape_owner_add_shape()` with the `owner_id` created from each child.

Then, to remove a group of shapes (removing a group children) use the `shape_owner_remove_shape` with each child's owner_id (which I assume would need to be stored somewhere). I suppose this could be encapsulated in some code add-on maybe.

Is that the guiding process?

**Comment by AndreaCatania** [comment]
Yes, thats sounds ok.
Bullet is really performant, though I doubt that its architecture allows you to achieve what you want. 

A voxel game should have million little boxes, colliding each other, if thats the case You should consider optimizing at least its memory model (I can expand this more if you want) though other solutions may be more effective.

**Comment by KevinTriplett** [comment]
Thanks for the quick reply -- in my use case, I want a KinematicBody (doer) to grab another KinematicBody (victim) and start whirling it around, smashing it into things. If the victim is a child of doer, it's automatically transformed by doer's rotations which is what I want. The problem was that victim collisions weren't happening.

**Comment by memphis88** [comment]
> You can create your own using the `CollisionObject` [API](https://docs.godotengine.org/en/stable/classes/class_collisionobject.html) API. Create a `shape_owner` using [`create_shape_owner()`](https://docs.godotengine.org/en/stable/classes/class_collisionobject.html#class-collisionobject-method-create-shape-owner), then add multiple shapes to it using [`shape_owner_add_shape()`](https://docs.godotengine.org/en/stable/classes/class_collisionobject.html#class-collisionobject-method-shape-owner-add-shape).

Thanks @madmiraal, that helped my use case which I couldn't find a solution for and I have a feeling it is somewhat common (I'm mostly referring to the limitation of shapes requiring to be direct children of PhysicsBody). To elaborate, I wanted to have my collision shape attached to a bone so that it follows the animation of my player in jumps etc, (see screenshot).
![image](https://user-images.githubusercontent.com/1369451/111881863-d7a48700-89b2-11eb-98a6-7297c0a62688.png)

**Comment by Shadowblitz16** [comment]
Can we get this for CollisionObject2D and CollisionObject3D? instead of RigidyBodys?

**Comment by aaronfranke** [comment]
@Shadowblitz16 Yes. That's what the PR I made does: https://github.com/godotengine/godot/pull/77937

**Comment by ctrlraul** [comment]
I have an use case where the amount of shapes is dynamic, they also need to be re-ordered, added, removed... Having a node to group for them would be extremely helpful as to not conflict with the other nodes in the rigid body (Like sprite, etc...)

**Comment by StoqnS14** [comment]
PLEASE ADD THIS PLEASE PLEASE PLEASE PLEASE I WANT THIS I NEED THIS

**Comment by TheMaydayMan** [comment]
This would be so very convenient in my polygons project where it's easier to have
RigidBody2D
|-- Polygon2D
      |-- CollisionPolygon2D

**Comment by mikest** [comment]
Hey, so there is a way this could be added and preserve backwards compatibility. The ownership search happens on "NOTIFICATION_PARENTED". That could be conditional to a new flag that says basically "search_ancestors". If the use wants a collision shape to include ancestors, they could check the checkbox and the collision shape would attach to the first ancestor in the tree that is a `CollisionObject3D`.

This would greatly simplify organizational tasks involving physics bodies.

Happy to do this work if I can find the maintainer for this part of the code base.

Edit:
Looks like @aaronfranke is the last person to touch this code.
